---
title: "R Notebook"
output: html_notebook
---

Names: Leon Kwan, Doron lisiansky, Richard Clarke
github repo: https://github.com/r-clarke/bioinformatics
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE119290

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if(!("ConsensusClusterPlus" %in% installed.packages()))
  BiocManager::install('ConsensusClusterPlus', update = FALSE)

if (!("Seurat" %in% installed.packages())) {
  BiocManager::install("Seurat", update = FALSE)
}
if (!("limma" %in% installed.packages())) {
  BiocManager::install("limma", update = FALSE)
}


```

```{r}

library(ConsensusClusterPlus)
library(Seurat)
library(limma)
library(dplyr)
library(patchwork)
library(clusterProfiler)
library(gprofiler2)
library(pheatmap)
library(RColorBrewer)
library(magrittr)
library(matrixStats)
library(ggplot2)
library(DESeq2)
```

```{r}
matrix_of_data <- as.matrix(read.table("GSE119290_Readhead_2018_RNAseq_gene_counts.txt"))
coldata<-read.csv("coldata.csv",header = T,row.names=1,stringsAsFactors=T)
coldata
```

```{r}
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:5000],]
d=sweep(d,1, apply(d,1,median,na.rm=T))

title="ConsensusClusterPlus"
results = ConsensusClusterPlus(d,maxK=3,reps=50,pItem=0.8,pFeature=1,title=title,clusterAlg="pam",distance="pearson")
icl = calcICL(results,title=title)
```


```{r}
for(gene_count in c(5000,10, 100,1000,10000)){
  mads=apply(matrix_of_data,1,mad)
  d=matrix_of_data[rev(order(mads))[1:gene_count],]
  d=sweep(d,1, apply(d,1,median,na.rm=T))
  title="ConsensusClusterPlus"
  results = ConsensusClusterPlus(d,maxK=3,reps=50,pItem=0.8,pFeature=1,title=title,clusterAlg="pam",distance="pearson")
}


```


```{r}
cts <- as.matrix(read.table("C:/Users/doron/source/semesterE-UF/Bioinformatics/bioinformatics/GSE119290_Readhead_2018_RNAseq_gene_counts.txt"))
coldata<-read.csv("C:/Users/doron/source/semesterE-UF/Bioinformatics/bioinformatics/coldata.csv",header = T,row.names=1,stringsAsFactors=T)

```

```{r}
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 5000)
head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.9)
head(Seurat::Idents(pbmc),42)
idents <- as.matrix(Seurat::Idents(pbmc))
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")
```

```{r}
pbmc.markers <- Seurat::FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>%
  group_by(cluster) %>%
  top_n(n = 2, wt = avg_log2FC)
top <- pbmc.markers %>%
  group_by(cluster) %>%
  top_n(n = 10, wt = avg_log2FC)

Seurat::DoHeatmap(pbmc, features = top$gene,group.bar = TRUE) 

colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
dds <- DESeqDataSetFromMatrix(countData = cts,colData = idents,design = ~ V1)
dds <- dds[rowSums(counts(dds)) > 1,]

vst <- vst(dds,blind = FALSE)
sampleDists <- dist(t(assay(vst)))
sampleDists


sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vst$V1, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
sampleDistMatrix
#Heatmap of sample-to-sample distances
pheatmap(sampleDistMatrix,clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,col = colors)

```

```{r}
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 10000)
head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.9)
head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")
```


```{r}
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 1000)
head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.9)
head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")
```

```{r}
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 100)
head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.9)
head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")
```


```{r}
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 10)
head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:5)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.9)
head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:5)
Seurat::DimPlot(pbmc, reduction = "umap")
```


```{r}
Seurat::Idents(pbmc)
coldata
chisq <- chisq.test(table(Seurat::Idents(pbmc), coldata$dex))
chisq$p.value
p.adjust(chisq$p.value,method="BH")
```