---
title: "R Notebook"
output: html_notebook
---

Names: Leon Kwan, Doron lisiansky, Richard Clarke
github repo: https://github.com/r-clarke/bioinformatics
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE119290

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if(!("ConsensusClusterPlus" %in% installed.packages()))
  BiocManager::install('ConsensusClusterPlus', update = FALSE)

if (!("Seurat" %in% installed.packages())) {
  BiocManager::install("Seurat", update = FALSE)
}
if (!("limma" %in% installed.packages())) {
  BiocManager::install("limma", update = FALSE)
}
if (!("ComplexHeatmap" %in% installed.packages())) {
  BiocManager::install("ComplexHeatmap", update = FALSE)
}

```

```{r}
library(cluster)
library(ComplexHeatmap)
library(dendextend)
library(ConsensusClusterPlus)
library(Seurat)
library(limma)
library(dplyr)
library(patchwork)
library(clusterProfiler)
library(gprofiler2)
library(pheatmap)
library(RColorBrewer)
library(magrittr)
library(matrixStats)
library(ggplot2)
library(DESeq2)
library(ggalluvial)
library(GGally)
library(broom)
```

```{r}
matrix_of_data <- as.matrix(read.table("GSE119290_Readhead_2018_RNAseq_gene_counts.txt"))
coldata<-read.csv("coldata.csv",header = T,row.names=1,stringsAsFactors=T)
```

```{r}
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:5000],]
d1=sweep(d,1, apply(d,1,median,na.rm=T))

title="ConsensusClusterPlus"
results1 = ConsensusClusterPlus(d1,maxK=6,reps=50,pItem=0.8,pFeature=1,title=title,clusterAlg="pam",distance="pearson")
icl = calcICL(results1,title=title)
```
Consensus clustering using PAM shows that k=2 has the best clustering stability, as shown in the heatmaps with two clean, blue, squares. The other k's are not as clearly clustered. Other plots show statistics related to consensus cdf and tracking plot. As shown in the item consensus bar graphs, the bars in k=2 were the most similar to each other in height.

```{r}
clust.row <- hclust(dist(d1), method="ward.D")
#clust.row <- pam(d1, k=2)
clust.col <- results1[[2]]$consensusTree
consensus_res <- results1[[2]]$consensusClass
consensus_res <- replace(consensus_res, consensus_res==1, "SZ")
consensus_res <- replace(consensus_res, consensus_res==2, "Control")
annotcol <- as.data.frame(consensus_res)

pheatmap(d1,
         #title="consensus cluster plus with PAM for 5000 genes",
         cluster_rows=clust.row,
         cluster_cols=clust.col,
         annotation_col = annotcol,
         show_rownames = FALSE,
         show_colnames = FALSE,
         scale = "row")
```
The heatmap for consensus clustering with PAM shows the differences in gene expression between control and schizophrenic patients. The control patients have a positive difference, as shown with more red, while, SZ patients have a negative difference in gene expression, as shown with more blue.

```{r}
gene_count = 10
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:gene_count],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
title="ConsensusClusterPlus"
results = ConsensusClusterPlus(d,maxK=6,reps=50,pItem=0.8,pFeature=1,title=title,clusterAlg="pam",distance="pearson", plot='pdf')
#ggplot(results[1], aes(y=Freq, axis1=))
consensus_res <- results[[2]]$consensusClass
consensus_res <- replace(consensus_res, consensus_res==1, "SZ")
consensus_res <- replace(consensus_res, consensus_res==2, "Control")
aluv_data = as.data.frame(table(names(consensus_res),consensus_res, coldata$dex))
ggplot(aluv_data,
     aes(y = Freq, axis1 = Var1, axis2 = consensus_res)) +
geom_alluvium(aes(fill = Var3), width = 1/12) +
geom_stratum(width = 1/12, fill = "black", color = "grey") +
geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
scale_fill_brewer(type = "qual", palette = "Set1") +
ggtitle("alluvial diagram for consensus clustering with 10 genes")
```
With only 10 genes, control ans SZ are clustered poorly. 

```{r}
gene_count = 100
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:gene_count],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
title="ConsensusClusterPlus"
results = ConsensusClusterPlus(d,maxK=6,reps=50,pItem=0.8,pFeature=1,title=title,clusterAlg="pam",distance="pearson", plot='pdf')
#ggplot(results[1], aes(y=Freq, axis1=))
consensus_res <- results[[2]]$consensusClass
consensus_res <- replace(consensus_res, consensus_res==1, "SZ")
consensus_res <- replace(consensus_res, consensus_res==2, "Control")
aluv_data = as.data.frame(table(names(consensus_res),consensus_res, coldata$dex))
ggplot(aluv_data,
     aes(y = Freq, axis1 = Var1, axis2 = consensus_res)) +
geom_alluvium(aes(fill = Var3), width = 1/12) +
geom_stratum(width = 1/12, fill = "black", color = "grey") +
geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
scale_fill_brewer(type = "qual", palette = "Set1") +
ggtitle("alluvial diagram for consensus clustering with 100 genes")
```
At 100 genes, control ans SZ are still clustered poorly. 


```{r}
gene_count = 1000
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:gene_count],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
title="ConsensusClusterPlus"
results = ConsensusClusterPlus(d,maxK=6,reps=50,pItem=0.8,pFeature=1,title=title,clusterAlg="pam",distance="pearson", plot='pdf')
#ggplot(results[1], aes(y=Freq, axis1=))
consensus_res <- results[[2]]$consensusClass
consensus_res <- replace(consensus_res, consensus_res==1, "SZ")
consensus_res <- replace(consensus_res, consensus_res==2, "Control")
aluv_data = as.data.frame(table(names(consensus_res),consensus_res, coldata$dex))
ggplot(aluv_data,
     aes(y = Freq, axis1 = Var1, axis2 = consensus_res)) +
geom_alluvium(aes(fill = Var3), width = 1/12) +
geom_stratum(width = 1/12, fill = "black", color = "grey") +
geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
scale_fill_brewer(type = "qual", palette = "Set1") +
ggtitle("alluvial diagram for consensus clustering with 1000 genes")
```
At 1000 genes, control ans SZ are still clustered poorly. 


```{r}
gene_count = 10000
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:gene_count],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
title="ConsensusClusterPlus"
results = ConsensusClusterPlus(d,maxK=6,reps=50,pItem=0.8,pFeature=1,title=title,clusterAlg="pam",distance="pearson", plot='pdf')
#ggplot(results[1], aes(y=Freq, axis1=))
consensus_res <- results[[2]]$consensusClass
consensus_res <- replace(consensus_res, consensus_res==1, "SZ")
consensus_res <- replace(consensus_res, consensus_res==2, "Control")
aluv_data = as.data.frame(table(names(consensus_res),consensus_res, coldata$dex))
ggplot(aluv_data,
     aes(y = Freq, axis1 = Var1, axis2 = consensus_res)) +
geom_alluvium(aes(fill = Var3), width = 1/12) +
geom_stratum(width = 1/12, fill = "black", color = "grey") +
geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
scale_fill_brewer(type = "qual", palette = "Set1") +
ggtitle("alluvial diagram for consensus clustering with 10,000 genes")
```
At 10000 genes, accuracy remains low.

```{r}
#coldata<-read.csv("coldata.csv",header = T,row.names=1,stringsAsFactors=T)
#levels(coldata$dex) <- c(levels(coldata$dex), 1)
#coldata$dex[coldata$dex == 'SZ'] <- 1
#levels(coldata$dex) <- c(levels(coldata$dex), 2)
#coldata$dex[coldata$dex == 'Control'] <- 2

consensus_res <- results1[[2]]$consensusClass
consensus_res <- replace(consensus_res, consensus_res==1, "SZ")
consensus_res <- replace(consensus_res, consensus_res==2, "Control")
chisq <- chisq.test(coldata$dex, consensus_res)
chisq
```
The high p-value of 0.7621 shows that the clustered values are not significantly different than the provided data. This means that this clustering method is well fitted to the sample data.

```{r}
cts <- as.matrix(read.table("GSE119290_Readhead_2018_RNAseq_gene_counts.txt"))
coldata<-read.csv("coldata.csv",header = T,row.names=1,stringsAsFactors=T)

```

Clustering Method: Hierarchical Clustering (hclust)
```{r}
dds <- DESeqDataSetFromMatrix(countData = matrix_of_data,colData = coldata,design = ~ dex)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#create distance matrix with the top 10 variable genes
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:10],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
d = scale(d)
di <- dist(t(d))

#perform hclustering and plot dendrogram
hc <- hclust(di, method="ward.D")
plot(hc, main = "Dendrogram for 10 Genes")

hc_res <- cutree(hclust(dist(t(d)), method="ward.D"), 2)
hc_res <- replace(hc_res, hc_res == 2, "SZ")
hc_res <- replace(hc_res, hc_res == 1, "Control")
aluv_data = as.data.frame(table(names(hc_res),hc_res, coldata$dex))
ggplot(aluv_data,
     aes(y = Freq, axis1 = Var1, axis2 = hc_res)) +
geom_alluvium(aes(fill = Var3), width = 1/12) +
geom_stratum(width = 1/12, fill = "black", color = "grey") +
geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
scale_fill_brewer(type = "qual", palette = "Set1") +
ggtitle("alluvial diagram for hierarchical clustering with 10 genes")

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#create distance matrix with the top 100 variable genes
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:100],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
d = scale(d)
di <- dist(t(d))

#perform hclustering and plot dendrogram
hc <- hclust(di, method="ward.D")
plot(hc, main = "Dendrogram for 100 Genes")

hc_res <- cutree(hclust(dist(t(d)), method="ward.D"), 2)
hc_res <- replace(hc_res, hc_res == 2, "SZ")
hc_res <- replace(hc_res, hc_res == 1, "Control")
aluv_data = as.data.frame(table(names(hc_res),hc_res, coldata$dex))
ggplot(aluv_data,
     aes(y = Freq, axis1 = Var1, axis2 = hc_res)) +
geom_alluvium(aes(fill = Var3), width = 1/12) +
geom_stratum(width = 1/12, fill = "black", color = "grey") +
geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
scale_fill_brewer(type = "qual", palette = "Set1") +
ggtitle("alluvial diagram for hierarchical clustering with 100 genes")

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#create distance matrix with the top 1000 variable genes
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:1000],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
d = scale(d)
di <- dist(t(d))

#perform hclustering and plot dendrogram
hc <- hclust(di, method="ward.D")
plot(hc, main = "Dendrogram for 1000 Genes")

hc_res <- cutree(hclust(dist(t(d)), method="ward.D"), 2)
hc_res <- replace(hc_res, hc_res == 2, "SZ")
hc_res <- replace(hc_res, hc_res == 1, "Control")
aluv_data = as.data.frame(table(names(hc_res),hc_res, coldata$dex))
ggplot(aluv_data,
     aes(y = Freq, axis1 = Var1, axis2 = hc_res)) +
geom_alluvium(aes(fill = Var3), width = 1/12) +
geom_stratum(width = 1/12, fill = "black", color = "grey") +
geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
scale_fill_brewer(type = "qual", palette = "Set1") +
ggtitle("alluvial diagram for hierarchical clustering with 1,000 genes")

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#create distance matrix with the top 10000 variable genes
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:10000],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
d = scale(d)
di <- dist(t(d))

#perform hclustering and plot dendrogram
hc <- hclust(di, method="ward.D")
plot(hc, main = "Dendrogram for 10000 Genes")

hc_res <- cutree(hclust(dist(t(d)), method="ward.D"), 2)
hc_res <- replace(hc_res, hc_res == 2, "SZ")
hc_res <- replace(hc_res, hc_res == 1, "Control")
aluv_data = as.data.frame(table(names(hc_res),hc_res, coldata$dex))
ggplot(aluv_data,
     aes(y = Freq, axis1 = Var1, axis2 = hc_res)) +
geom_alluvium(aes(fill = Var3), width = 1/12) +
geom_stratum(width = 1/12, fill = "black", color = "grey") +
geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
scale_fill_brewer(type = "qual", palette = "Set1") +
ggtitle("alluvial diagram for hierarchical clustering with 10,000 genes")

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#create distance matrix with the top 5000 variable genes
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:5000],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
d = scale(d)
di <- dist(t(d))

#perform hclustering and plot dendrogram
hc <- hclust(di, method="ward.D")
plot(hc, main = "Dendrogram for 5000 Genes")

hc_res <- cutree(hclust(dist(t(d)), method="ward.D"), 2)
hc_res <- replace(hc_res, hc_res == 2, "SZ")
hc_res <- replace(hc_res, hc_res == 1, "Control")

aluv_data = as.data.frame(table(names(hc_res),hc_res, coldata$dex))
ggplot(aluv_data,
     aes(y = Freq, axis1 = Var1, axis2 = hc_res)) +
geom_alluvium(aes(fill = Var3), width = 1/12) +
geom_stratum(width = 1/12, fill = "black", color = "grey") +
geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
scale_fill_brewer(type = "qual", palette = "Set1") +
ggtitle("alluvial diagram for hierarchical clustering with 5,000 genes")

```
After running hclust, we are able to view the dendrograms of the hierarchical clustering method. Looking first at the results from the top 5000 variable genes, it is clear that the dendrogram splits into 2 main clusters. This is why the k value 2 was chosen for the clustering in the heatmap below. Changing the k value created more clusters but left 2 with majority of the genes and the rest with either 0 genes or very few in the cluster.

Comparing to the results of running the clustering method with 10, 100, 1000, and 10000 genes, all of the results show 2 main groups or clusters in the dendrogram of results. A smaller cluster that splits off from a much larger cluster in each of the results from the clusting method. 

Looking at the alluvial diagrams, only for the clustering method using 100 genes did we notice a much larger mismatch of SZ and Control values. The rest of the diagrams are mostly correct with their clustering of SZ and Control samples.

```{r}
#draw heatmap with clusters and dendrograms
row_dend = as.dendrogram(hclust(dist(d), method="ward.D"))# row clustering
row_dend = color_branches(row_dend, k = 2)
col_dend = as.dendrogram(hclust(dist(t(d)), method="ward.D")) # column clustering
col_dend = color_branches(col_dend, k = 2)

suppressMessages(
Heatmap(
          d, 
          name = "expression", 
          column_title = c("Cluster 1", "Cluster 2"),
          row_title = c("Cluster 1", "Cluster 2"),
          row_title_rot = 0,
          row_names_gp = gpar(fontsize = 7),
          cluster_rows = row_dend,
          cluster_columns = col_dend,
          row_split = 2,
          column_split = 2
        )
)

```
Chi-Squared portion for Hierarchical Clustering
```{r}
sample_cluster <- hclust(dist(t(d)), method="ward.D")
hclust_results <- cutree(sample_cluster, k = 2)
hclust_results <- replace(hclust_results, hclust_results == 2, "SZ")
hclust_results <- replace(hclust_results, hclust_results == 1, "Control")

chisq <- chisq.test(coldata$dex, hclust_results)
chisq


```
With 1 degree of freedom and a X-squared value of 0.48485, differences between the expected and observed numbers occur my change between 25% and 50% of the time after looking at the Chi Squared distribution table. With this evaluation, we fail to reject our hypothesis.

```{r}
cts <- as.matrix(read.table("GSE119290_Readhead_2018_RNAseq_gene_counts.txt"))
coldata<-read.csv("coldata.csv",header = T,row.names=1,stringsAsFactors=T)

```


```{r}
# performing knn clustring method with 5000 genes.
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
#pbmc
head(pbmc@meta.data, 5)
### Normalizing the data
#We will normalize the data ensuring expression values across cells are on a comparable scale
pbmc <- Seurat::NormalizeData(pbmc)
# Identify the 5000 most highly variable genes
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 5000)
head(pbmc@assays, 5)
### Data scaling
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
### Linear dimensional reduction
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")

### Determine the dimentionality of dataset
Seurat::ElbowPlot(pbmc)
### Cluster the data
#K-nearest neighbor (KNN) graph based on the euclidean distance in PCA space
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)
pbmc <- Seurat::FindClusters(pbmc, resolution = 0.95)
#head(Seurat::Idents(pbmc),42)
#saving the final classes from the clustering
idents <- as.matrix(Seurat::Idents(pbmc))
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")

knn_res <- Seurat::Idents(pbmc)

```

```{r}
# find markers for every cluster compared to all remaining samples, report only the positive ones
pbmc.markers <- Seurat::FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>%
  group_by(cluster) %>%
  top_n(n = 2, wt = avg_log2FC)
top <- pbmc.markers %>%
  group_by(cluster) %>%
  top_n(n = 10, wt = avg_log2FC)

#draw heatmap with clusters and dendrograms

Seurat::DoHeatmap(pbmc, features = top$gene,group.bar = TRUE) 

colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
dds <- DESeqDataSetFromMatrix(countData = cts,colData = idents,design = ~ V1)
dds <- dds[rowSums(counts(dds)) > 1,]

vst <- vst(dds,blind = FALSE)
sampleDists <- dist(t(assay(vst)))
#sampleDists


sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vst$V1, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Reds")) )(255)
#sampleDistMatrix
#Heatmap of sample-to-sample distances
pheatmap(sampleDistMatrix,clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,col = colors)
topVarGenes <- head(order(rowVars(assay(vst)), decreasing = TRUE), 5000)
#topVarGenes
mat  <- assay(vst)[topVarGenes, ]

mat<-mat-rowMeans(mat)
#mat
anno <- as.data.frame(colData(vst)[c("V1")])
pheatmap(mat, annotation_col = anno)

#Create an alluvial diagram 

aluv_data = as.data.frame(table(Seurat::Cells(pbmc),Seurat::Idents(pbmc), coldata$dex))
#aluv_data
ggplot(aluv_data,
       aes(y = Freq, axis1 = Var1, axis2 = Var2)) +
  geom_alluvium(aes(fill = Var3), width = 1/12) +
  geom_stratum(width = 1/12, fill = "black", color = "grey") +
  geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
  scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
  scale_fill_brewer(type = "qual", palette = "Set1") +
  ggtitle("alluvial diagram for knn with 5000 genes")

#adding chi squared results for final statistics
chisq <- chisq.test(table(Seurat::Idents(pbmc), coldata$dex))
chisq
data <- data.frame(knn_chi_5000 = chisq$p.value)
#data 
#p.adjust(chisq$p.value,method="BH")
```

Seurat clusters cells based on their PCA scores in order to overcome the extensive technical noise in any single feature for scRNA-seq data. The top principal components represent a robust compression of the dataset.

To decide how many components should we choose to include, we applied the Elbow plot method: a ranking of principle components based on the percentage of variance explained by each one.

first we ran knn with the top 5000 variable genes, we got 2 clasess and from the alluvial plot we can see the clustering was very good in compare to the real groups(sz, control)

in this specific method we are not selecting  k, Setting the resolution parameter is changing the number of clusters , from what i read , resolution between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells.



```{r}
# performing knn clustring method with 10000 genes.
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
#pbmc
#head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 10000)
#head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
#pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)

pbmc <- Seurat::FindClusters(pbmc, resolution = 1.)
#head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")
    
aluv_data = as.data.frame(table(Seurat::Cells(pbmc),Seurat::Idents(pbmc), coldata$dex))
  #aluv_data
  ggplot(aluv_data,
       aes(y = Freq, axis1 = Var1, axis2 = Var2)) +
  geom_alluvium(aes(fill = Var3), width = 1/12) +
  geom_stratum(width = 1/12, fill = "black", color = "grey") +
  geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
  scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
  scale_fill_brewer(type = "qual", palette = "Set1") +
  ggtitle("alluvial diagram for knn with 10000 genes")

chisq <- chisq.test(table(Seurat::Idents(pbmc), coldata$dex))
chisq$p.value
data$knn_chi_10000 = chisq$p.value  
#p.adjust(chisq$p.value,method="BH")  
```
performing knn clustring method with 10000 genes.
we Initialize the Seurat object with the data , then We will normalize the data ensuring expression values across cells are on a comparable scale.
We next calculate a subset of features that exhibit high gene-to-gene variation in the dataset , we will scale the data and run pca.
we will cluster the data using K-nearest neighbor (KNN) graph based on the euclidean distance in PCA space and will set the resolution parameter(which affects on the number of clusters).
in the end we will plot the alluvial plot and compute the chi-squared test.

```{r}
# performing knn clustring method with 1000 genes.
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
#pbmc
#head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 1000)
#head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
#pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.95)
#head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")

aluv_data = as.data.frame(table(Seurat::Cells(pbmc),Seurat::Idents(pbmc), coldata$dex))
  #aluv_data
  ggplot(aluv_data,
       aes(y = Freq, axis1 = Var1, axis2 = Var2)) +
  geom_alluvium(aes(fill = Var3), width = 1/12) +
  geom_stratum(width = 1/12, fill = "black", color = "grey") +
  geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
  scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
  scale_fill_brewer(type = "qual", palette = "Set1") +
  ggtitle("alluvial diagram for knn with 1000 genes")
  
chisq <- chisq.test(table(Seurat::Idents(pbmc), coldata$dex))
chisq$p.value
data$knn_chi_1000 = chisq$p.value   
```
performing knn clustring method with 1000 genes.
we Initialize the Seurat object with the data , then We will normalize the data ensuring expression values across cells are on a comparable scale.
We next calculate a subset of features that exhibit high gene-to-gene variation in the dataset , we will scale the data and run pca.
we will cluster the data using K-nearest neighbor (KNN) graph based on the euclidean distance in PCA space and will set the resolution parameter(which affects on the number of clusters).
in the end we will plot the alluvial plot and compute the chi-squared test.

```{r}
# performing knn clustring method with 100 genes.
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
#pbmc
#head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 100)
#head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
#pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.95)
#head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")

aluv_data = as.data.frame(table(Seurat::Cells(pbmc),Seurat::Idents(pbmc), coldata$dex))
  #aluv_data
  ggplot(aluv_data,
       aes(y = Freq, axis1 = Var1, axis2 = Var2)) +
  geom_alluvium(aes(fill = Var3), width = 1/12) +
  geom_stratum(width = 1/12, fill = "black", color = "grey") +
  geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
  scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
  scale_fill_brewer(type = "qual", palette = "Set1") +
  ggtitle("alluvial diagram for knn with 100 genes")
  
chisq <- chisq.test(table(Seurat::Idents(pbmc), coldata$dex))
chisq$p.value
data$knn_chi_100 = chisq$p.value   
```
performing knn clustring method with 100 genes.
we Initialize the Seurat object with the data , then We will normalize the data ensuring expression values across cells are on a comparable scale.
We next calculate a subset of features that exhibit high gene-to-gene variation in the dataset , we will scale the data and run pca.
we will cluster the data using K-nearest neighbor (KNN) graph based on the euclidean distance in PCA space and will set the resolution parameter(which affects on the number of clusters).
in the end we will plot the alluvial plot and compute the chi-squared test.

```{r}
# performing knn clustring method with 10 genes.
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
#pbmc
#head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 10)
#head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
#pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:8)

pbmc <- Seurat::FindClusters(pbmc, resolution =0.95)
#head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:8)
Seurat::DimPlot(pbmc, reduction = "umap")

aluv_data = as.data.frame(table(Seurat::Cells(pbmc),Seurat::Idents(pbmc), coldata$dex))
  #aluv_data
  ggplot(aluv_data,
       aes(y = Freq, axis1 = Var1, axis2 = Var2)) +
  geom_alluvium(aes(fill = Var3), width = 1/12) +
  geom_stratum(width = 1/12, fill = "black", color = "grey") +
  geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
  scale_x_discrete(limits = c("samples", "cluster groups"), expand = c(.05, .05)) +
  scale_fill_brewer(type = "qual", palette = "Set1") +
  ggtitle("alluvial diagram for knn with 10 genes")
  
chisq <- chisq.test(table(Seurat::Idents(pbmc), coldata$dex))
chisq$p.value
data$knn_chi_10 = chisq$p.value   
```
Comparing to the results of running the clustering method with 10, 100, 1000, and 10000 genes, all of the results show 2/3 clusters . the results for each number off genes is a bit different even if it return the same number of clusters and we can see in the alluvial plots how they are accurate.


```{r}

pairs <- as.data.frame(coldata)
colnames(pairs)[1] <- "samples"
pairs$knn <- knn_res
pairs$consensus <- consensus_res
pairs$hc <- hc_res

all_statistical_results <- data.frame("KNN_results" = c(1,1,0), "consensus_results" = c(1,1,0), "hclust_results" = c(1,1,0))
rownames(all_statistical_results) <- c("chisq value", "Original P value", "Adjusted P Value")

chisq <- chisq.test(coldata$dex, pairs$hc)
all_statistical_results$hclust_results <- c(chisq$statistic, chisq$p.value, 0)

chisq <- chisq.test(coldata$dex, pairs$consensus)
all_statistical_results$consensus_results <- c(chisq$statistic, chisq$p.value, 0)

chisq <- chisq.test(coldata$dex, pairs$knn)

all_statistical_results$KNN_results <- c(chisq$statistic, chisq$p.value, 0)
tempPvalues <- p.adjust(all_statistical_results[2,1:3],method="BH")
all_statistical_results[3,1] <- tempPvalues[1]
all_statistical_results[3,2] <- tempPvalues[2]
all_statistical_results[3,3] <- tempPvalues[3]
all_statistical_results
```
From the results, the consensus clustering with PAM algorithm and hierarchical clustering algorithm were able to group had high p-values, meaning they were not statistically different than the sample label. This means their clustering fits the sample data well. Meanwhile, the KNN algorithm clustered the data significantly different than the sample with a p-value < 0.05, which means it did not fit the data well.