---
title: "R Notebook"
output: html_notebook
---

Names: Leon Kwan, Doron lisiansky, Richard Clarke
github repo: https://github.com/r-clarke/bioinformatics
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE119290

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if(!("ConsensusClusterPlus" %in% installed.packages()))
  BiocManager::install('ConsensusClusterPlus', update = FALSE)

if(!("ComplexHeatmap" %in% installed.packages()))
  BiocManager::install('ComplexHeatmap', update = FALSE)

if (!("Seurat" %in% installed.packages())) {
  BiocManager::install("Seurat", update = FALSE)
}
if (!("limma" %in% installed.packages())) {
  BiocManager::install("limma", update = FALSE)
}


```

```{r}

library(ConsensusClusterPlus)
library(Seurat)
library(limma)
library(dplyr)
library(patchwork)
library(clusterProfiler)
library(gprofiler2)
library(pheatmap)
library(RColorBrewer)
library(magrittr)
library(matrixStats)
library(ggplot2)
library(DESeq2)
library(dendextend)
library(ComplexHeatmap)
```

```{r}
matrix_of_data <- as.matrix(read.table("GSE119290_Readhead_2018_RNAseq_gene_counts.txt"))
coldata<-read.csv("coldata.csv",header = T,row.names=1,stringsAsFactors=T)
coldata
```

```{r}
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:5000],]
d1=sweep(d,1, apply(d,1,median,na.rm=T))

title="ConsensusClusterPlus"
results = ConsensusClusterPlus(d1,maxK=6,reps=50,pItem=0.8,pFeature=1,title=title,clusterAlg="pam",distance="pearson")
icl = calcICL(results,title=title)
```



```{r}
for(gene_count in c(10, 100,1000,10000)){
  mads=apply(matrix_of_data,1,mad)
  d=matrix_of_data[rev(order(mads))[1:gene_count],]
  d=sweep(d,1, apply(d,1,median,na.rm=T))
  title="ConsensusClusterPlus"
  results = ConsensusClusterPlus(d,maxK=6,reps=50,pItem=0.8,pFeature=1,title=title,clusterAlg="pam",distance="pearson")
  #ggplot(results[1], aes(y=Freq, axis1=))
}


```
```{r}
## notice the use of a variable (rather then using 3 directly in the following commands)
nc <- 2

annot <- read.table("GSE119290_Readhead_2018_RNAseq_gene_counts.txt")
rownames(annot) <- colnames(assayData)
colnames(annot) <- c("DiseaseState")

## remake heatmap, include both subtype and cluster assignments for visual comparison
annot1 <- data.frame(annot,cluster=results[[nc]]$consensusClass)
annotCol$cluster <- rainbow(n=nc)
names(annotCol$cluster) <- unique(annot1$cluster)

## use cluster tree from consensus clustering for column ordering in heatmap
clust.col <- results[[nc]]$consensusTree
## determine row ordering based on de-novo clustering
clust.row <- hcopt(as.dist(1-cor(t(exprs(matrix_of_data)))),method="ward.D")

##featureNames(cancerSet) <- fData(cancerSet)$hgnc_symbol
heatmaptitle <- "Heatmap consensus clustering assignment vs. subtype\n top 4k z-score normalized"
pheatmap(d1,
         title=heatmaptitle,
         #annotation_col = annot,
         #annotation_colors = annotCol,
         #cluster_rows=clust.row,
         cluster_cols=clust.col,
         show_rownames = FALSE,
         show_colnames = FALSE,
         scale = "row")
```



Clustering Method: Hierarchical Clustering (hclust)
```{r}
dds <- DESeqDataSetFromMatrix(countData = matrix_of_data,colData = coldata,design = ~ dex)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#create distance matrix with the top 10 variable genes
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:10],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
d = scale(d)
di <- dist(d)

#perform hclustering and plot dendrogram
hc <- hclust(di, method="ward.D")
plot(hc)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#create distance matrix with the top 100 variable genes
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:100],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
d = scale(d)
di <- dist(d)

#perform hclustering and plot dendrogram
hc <- hclust(di, method="ward.D")
plot(hc)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#create distance matrix with the top 1000 variable genes
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:1000],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
d = scale(d)
di <- dist(d)

#perform hclustering and plot dendrogram
hc <- hclust(di, method="ward.D")
plot(hc)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#create distance matrix with the top 10000 variable genes
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:10000],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
d = scale(d)
di <- dist(d)

#perform hclustering and plot dendrogram
hc <- hclust(di, method="ward.D")
plot(hc)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#create distance matrix with the top 5000 variable genes
mads=apply(matrix_of_data,1,mad)
d=matrix_of_data[rev(order(mads))[1:5000],]
d=sweep(d,1, apply(d,1,median,na.rm=T))
d = scale(d)
di <- dist(d)

#perform hclustering and plot dendrogram
hc <- hclust(di, method="ward.D")
plot(hc)
```
After running hclust, we are able to view the dendrograms of the hierarchical clustering method. Looking first at the results from the top 5000 variable genes, it is clear that the dendrogram splits into 2 main clusters. This is why the k value 2 was chosen for the clustering in the heatmap below. Changing the k value created more clusters but left 2 with majority of the genes and the rest with either 0 genes or very few in the cluster.

Comparing to the results of running the clustering method with 10, 100, 1000, and 10000 genes, all of the results show 2 main groups or clusters in the dendrogram of results. A smaller cluster that splits off from a much larger cluster in each of the results from the clusting method. 

```{r}
#draw heatmap with clusters and dendrograms
row_dend = as.dendrogram(hclust(dist(d), method="ward.D"))# row clustering
row_dend = color_branches(row_dend, k = 2)
col_dend = as.dendrogram(hclust(dist(t(d)), method="ward.D")) # column clustering
col_dend = color_branches(col_dend, k = 2)

suppressMessages(
Heatmap(
          d, 
          name = "expression", 
          column_title = c("Cluster 1", "Cluster 2"),
          row_title = c("Cluster 1", "Cluster 2"),
          row_title_rot = 0,
          row_names_gp = gpar(fontsize = 7),
          cluster_rows = row_dend,
          cluster_columns = col_dend,
          row_split = 2,
          column_split = 2
        )
)

```
Chi-Squared portion for Hierarchical Clustering
```{r}

```


```{r}
cts <- as.matrix(read.table("C:/Users/doron/source/semesterE-UF/Bioinformatics/bioinformatics/GSE119290_Readhead_2018_RNAseq_gene_counts.txt"))
coldata<-read.csv("C:/Users/doron/source/semesterE-UF/Bioinformatics/bioinformatics/coldata.csv",header = T,row.names=1,stringsAsFactors=T)

```

```{r}
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 5000)
head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.9)
head(Seurat::Idents(pbmc),42)
idents <- as.matrix(Seurat::Idents(pbmc))
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")
```

```{r}
pbmc.markers <- Seurat::FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>%
  group_by(cluster) %>%
  top_n(n = 2, wt = avg_log2FC)
top <- pbmc.markers %>%
  group_by(cluster) %>%
  top_n(n = 10, wt = avg_log2FC)

Seurat::DoHeatmap(pbmc, features = top$gene,group.bar = TRUE) 

colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
dds <- DESeqDataSetFromMatrix(countData = cts,colData = idents,design = ~ V1)
dds <- dds[rowSums(counts(dds)) > 1,]

vst <- vst(dds,blind = FALSE)
sampleDists <- dist(t(assay(vst)))
sampleDists


sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vst$V1, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
sampleDistMatrix
#Heatmap of sample-to-sample distances
pheatmap(sampleDistMatrix,clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,col = colors)

```

```{r}
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 10000)
head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.9)
head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")
```


```{r}
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 1000)
head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.9)
head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")
```

```{r}
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 100)
head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:20)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.9)
head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:20)
Seurat::DimPlot(pbmc, reduction = "umap")
```


```{r}
pbmc <- Seurat::CreateSeuratObject(counts = cts, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
head(pbmc@meta.data, 5)
pbmc <- Seurat::NormalizeData(pbmc)
pbmc <- Seurat::FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 10)
head(pbmc@assays, 5)
all.genes <- rownames(pbmc)
pbmc <- Seurat::ScaleData(pbmc, features = all.genes)
pbmc <- Seurat::RunPCA(pbmc, features = Seurat::VariableFeatures(pbmc), npcs =43)
Seurat::DimPlot(pbmc, reduction = "pca")
pbmc
Seurat::ElbowPlot(pbmc)
pbmc <- Seurat::FindNeighbors(pbmc, dims = 1:5)

pbmc <- Seurat::FindClusters(pbmc, resolution = 0.9)
head(Seurat::Idents(pbmc),42)
pbmc <- Seurat::RunUMAP(pbmc, dims = 1:5)
Seurat::DimPlot(pbmc, reduction = "umap")
```


```{r}
Seurat::Idents(pbmc)
coldata
chisq <- chisq.test(table(Seurat::Idents(pbmc), coldata$dex))
chisq$p.value
p.adjust(chisq$p.value,method="BH")
```