---
title: "Tutorial_11_Markdown"
date: "6/21/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---
The depmap package aims to provide a reproducible research framework to cancer dependency data. The data found in the depmap package has been formatted to facilitate the use of common R packages such as dplyr and ggplot2. This package allow researchers to more easily mine, explore and visually illustrate dependency data taken from the Depmap cancer genomic dependency study.

```{r message=FALSE, warning=FALSE}
library("dplyr")
library("ggplot2")
library("viridis")
library("tibble")
library("gridExtra")
library("stringr")
library("depmap")
library("ExperimentHub") 
```


```{r message=FALSE, warning=FALSE}
#ExperimentHub is a client for the BioconductorExperimentHub web resource. ExperimentHub provides a central location where curated data from experiments, publications or training courses can be accessed.
eh <- ExperimentHub()
#mcols(eh)
query(eh, "depmap")
```


By importing the depmap data into the R environment, the data can be mined more effectively. Specific datasets can be accessed by EH number.
```{r message=FALSE, warning=FALSE}
rnai <- eh[["EH3080"]]
mutationCalls <- eh[["EH3085"]]
metadata <- eh[["EH3086"]]
TPM <- eh[["EH3084"]]
copyNumber <- eh[["EH3082"]]
```

For example, if one interested researching soft tissue sarcomas and wanted to search all such cancer cell lines for the gene with the greatest dependency, it is possible to accomplish this task by using data manipulation and visualization tools dplyr and ggplot2. 

Below, the rnai dataset is selected for cell lines with “SOFT_TISSUE” in their Broad Institute Cancer Cell Line Encyclopedia name (all the CCLE names can be found [here](https://portals.broadinstitute.org/ccle/about) ), and displaying a list of the highest dependency scores.

```{r}
## list of dependency scores
rnai %>% dplyr::select(cell_line, gene_name, dependency) %>%
         dplyr::filter(stringr::str_detect(cell_line, "SOFT_TISSUE")) %>%
         dplyr::arrange(dependency) %>% 
         head(10)
```

As the gene RPL14 appears several times in the top dependency scores, it may make an interesting candidate target. Below, a plot of the rnai data is displayed as a histogram showing the distribution of dependency scores for gene RPL14.

```{r}

## We build a histogram of dependency scores
rnai %>% dplyr::select(gene_name, dependency) %>% 
         dplyr::filter(gene_name == "RPL14") %>% 
         ggplot(aes(x = dependency)) +
         geom_histogram() +
         geom_vline(xintercept = mean(rnai$dependency, na.rm = TRUE),
                    linetype = "dotted", color = "red") +
         ggtitle("Histogram of dependency scores for gene RPL14")
```

A more complex plot of the rnai data involves plotting the distribution of dependency scores for gene RPL14 for each major type of cancer, while highlighting the nature of mutations of this gene in such cancer cell lines (e.g. if such are COSMIC hotspots, damaging, etc.).


```{r}
head(metadata) #cell lines metadata
head(rnai)
head(mutationCalls)

meta_rnai <- metadata %>%
             dplyr::select(depmap_id, lineage) %>%
             dplyr::full_join(rnai, by = "depmap_id") %>%
             dplyr::filter(gene_name == "RPL14") %>% 
             dplyr::full_join((mutationCalls %>%
                              dplyr::select(depmap_id, entrez_id,
                                            is_cosmic_hotspot, var_annotation)),
                                            by = c("depmap_id", "entrez_id"))

p1 <- meta_rnai %>%
      ggplot(aes(x = dependency, y = lineage)) +
      geom_point(alpha = 0.4, size = 0.5) +
      geom_point(data = subset(
         meta_rnai, var_annotation == "damaging"), color = "red") +
      geom_point(data = subset(
         meta_rnai, var_annotation == "other non-conserving"), color = "blue") +
      geom_point(data = subset(
         meta_rnai, var_annotation == "other conserving"), color = "cyan") +
      geom_point(data = subset(
         meta_rnai, is_cosmic_hotspot == TRUE), color = "orange") +
      ggtitle("Scatterplot of dependency scores for gene RPL14 by lineage")

p1
```

Below is a boxplot displaying expression values for gene RPL14 by lineage:

```{r}
metadata %>%
      dplyr::select(depmap_id, lineage) %>%
      dplyr::full_join(TPM, by = "depmap_id") %>%
      dplyr::filter(gene_name == "RPL14") %>% 
      ggplot(aes(x = lineage, y = expression, fill = lineage)) +
      geom_boxplot(outlier.alpha = 0.1) +
      ggtitle("Boxplot of expression values for gene RPL14 by lineage") +
      theme(axis.text.x = element_text(angle = 45, hjust=1)) +
      theme(legend.position = "none")

```


Let's look for high dependency and high expression genes that can be interesting research targets. 
Let's build a plot of expression vs rnai gene dependency for Rhabdomyosarcoma:


```{r}
## expression vs rnai gene dependency for Rhabdomyosarcoma:
sarcoma <- metadata %>%
           dplyr::select(depmap_id, cell_line,
                         primary_disease, subtype_disease) %>%
           dplyr::filter(primary_disease == "Sarcoma",
                         subtype_disease == "Rhabdomyosarcoma")

rnai_sub <- rnai %>% dplyr::select(depmap_id, gene, gene_name, dependency)
tpm_sub <- TPM %>% dplyr::select(depmap_id, gene, gene_name, expression)

sarcoma_dep <- sarcoma %>%
               dplyr::left_join(rnai_sub, by = "depmap_id") %>%
               dplyr::select(-cell_line, -primary_disease,
                             -subtype_disease, -gene_name)

sarcoma_exp <- sarcoma %>% dplyr::left_join(tpm_sub, by = "depmap_id")

sarcoma_dat_exp <- dplyr::full_join(sarcoma_dep, sarcoma_exp,
                             by = c("depmap_id", "gene")) %>%
                             dplyr::filter(!is.na(expression))

p2 <- ggplot(data = sarcoma_dat_exp, aes(x = dependency, y = expression)) +
      geom_point(alpha = 0.4, size = 0.5) +
      geom_vline(xintercept=mean(sarcoma_dat_exp$dependency, na.rm = TRUE),
                 linetype = "dotted", color = "red") +
      geom_hline(yintercept=mean(sarcoma_dat_exp$expression, na.rm = TRUE),
                 linetype = "dotted", color = "red") +
      ggtitle("Scatterplot of rnai dependency vs expression values for gene")
p2 + theme(axis.text.x = element_text(angle = 45))
```


A selection of the genes shown above with the lowest dependency scores, also displaying gene expression in TPM in the last column.

```{r}
sarcoma_dat_exp %>%
    dplyr::select(cell_line, gene_name, dependency, expression) %>%
    dplyr::arrange(dependency) %>% 
    head(10)
```


